### PR TITLE
Add integration tests for Organizations Discovery GET API

### PR DESCRIPTION
**Pull request by BimsaraBodaragama**
### Description

> This PR introduces a set of comprehensive integration tests for the Organization Discovery GET API, focusing on validating pagination and cursor-based navigation. The tests also cover various edge cases and failure scenarios, ensuring that the API responds correctly to invalid inputs and handles different pagination limits and offsets.

### Purpose

> The purpose of this PR is to enhance the test coverage of the Organization Discovery GET API by implementing thorough tests that verify the correct functionality of pagination, including both forward and backward navigation, and by ensuring the API's robustness against invalid inputs.

### Goals

> - Verify that the pagination and offset-based navigation features in the Organization Discovery GET API function are as expected across various scenarios.
> - To ensure the API returns appropriate error responses when invalid pagination parameters, such as incorrect limits and offsets, are provided.
> - To improve the overall reliability of the API by covering standard use cases, as well as numeric and non-numeric edge cases.

### Approach

> 1. Success Tests:
> - Implemented tests for enabling email domain discovery and added email domains to multiple organizations for subsequent pagination tests.
> - Created a series of tests that validate pagination with different limits (e.g., 3, 5, 10, 15, 17, 20, 25), ensuring that the API correctly returns organizations based on the provided limit.
> - Added tests for offset-based pagination, validating the correct behavior of navigating forward and backward through organization pages, with checks for both `next` and `previous` links.
> - Included edge case tests for limits and offsets, verifying the API's behavior when limits are zero or exceed the total number of organizations.
>
> 2. Failure Tests:
> - Added tests to ensure that providing an invalid pagination limit or offset (e.g., negative values, zero, or missing values) results in appropriate error responses, such as 400 Bad Request Error.
> - Implemented scenarios that validate the API's handling of non-numeric edge cases, such as empty or missing limit and offset parameters, ensuring robust error handling.

### CODE REVIEW

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/organization/management/v1/OrganizationManagementFailureTest.java [hunk 1]
```diff
@@ -380,26 +381,77 @@ public void deleteOrganizations() {
     @Test(dependsOnMethods = "testUpdateDiscoveryAttributesUnauthorized")
     public void testGetPaginatedOrganizationsWithInvalidLimit() {
 
-        String invalidLimitUrl = ORGANIZATION_MANAGEMENT_API_BASE_PATH + QUESTION_MARK + LIMIT_QUERY_PARAM + EQUAL + "-1";
+        String invalidLimitUrl =
```
[the remark this answers is not in the document]
Comment (ShanChathusanda93) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/organization/management/v1/OrganizationManagementFailureTest.java:384]: Please check the other places as well.

[the remark this answers is not in the document]
Comment (BimsaraBodaragama) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/organization/management/v1/OrganizationManagementFailureTest.java:384]: It happens when formating due to the line wrap at 120

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/organization/management/v1/OrganizationManagementFailureTest.java [hunk 2]
```diff
@@ -380,26 +381,77 @@ public void deleteOrganizations() {
...
+                        + AMPERSAND + BEFORE_QUERY_PARAM + EQUAL + INVALID_CURSOR;
         Response response = getResponseOfGetWithOAuth2(invalidBeforeCursorUrl, m2mToken);
         validateErrorResponse(response, HttpStatus.SC_BAD_REQUEST, ERROR_CODE_INVALID_PAGINATION_CURSOR);
     }
+
+    @DataProvider(name = "organizationDiscoveryInvalidLimitAndOffsetDataProvider")
+    public Object[][] organizationDiscoveryInvalidLimitAndInvalidOffsetDataProvider() {
+
+        return new Object[][]{
+                {"0", "-1"}, {"5", "-1"}, {"20", "-1"}, {"25", "-1"}, {"", "-1"}, //invalid limit
```
Comment (sahandilshan) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/organization/management/v1/OrganizationManagementFailureTest.java:414]:
```suggestion
                {"0", "-1"}, {"5", "-1"}, {"20", "-1"}, {"25", "-1"}, {"", "-1"}, // Invalid limit.
```

Comment (sahandilshan) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/organization/management/v1/OrganizationManagementFailureTest.java:414]: Comment should be in following format
1. First letter should be capital
2. Ends with full stop

Do the change in other places as well

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/organization/management/v1/OrganizationManagementSuccessTest.java [hunk 3]
```diff
@@ -785,48 +786,382 @@ public void testGetPaginatedOrganizationsWithDefaultLimit() {
...
+
+        // Test case 2: URL without LIMIT_QUERY_PARAM.
+        String endpointURLWithoutLimit =
+                ORGANIZATION_MANAGEMENT_API_BASE_PATH + ORGANIZATION_DISCOVERY_API_PATH + QUESTION_MARK +
+                        FILTER_QUERY_PARAM + EQUAL + AMPERSAND + OFFSET_QUERY_PARAM + EQUAL + ZERO;
+
+        Response responseWithoutLimit = getResponseOfGetWithOAuth2(endpointURLWithoutLimit, m2mToken);
+        validateHttpStatusCode(responseWithoutLimit, HttpStatus.SC_OK);
+
+        validateResponseForOrganizationDiscoveryLimitDefaultCases(responseWithoutLimit);
```
Comment (sahandilshan) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/organization/management/v1/OrganizationManagementSuccessTest.java:989]: You can split this into two test methods

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/organization/management/v1/OrganizationManagementFailureTest.java (hunk 2 above)
Comment (BimsaraBodaragama) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/organization/management/v1/OrganizationManagementFailureTest.java]: Addressed with commit: https://github.com/wso2/product-is/pull/21023/commits/7629f4c85aa7707952c2962016ec9bcd46e868a8

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/organization/management/v1/OrganizationManagementSuccessTest.java (hunk 3 above)
Comment (BimsaraBodaragama) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/organization/management/v1/OrganizationManagementSuccessTest.java]: Addressed with commit: https://github.com/wso2/product-is/pull/21023/commits/6ce4d2238cbc4e98ef8071dd6fa6ec098d6f9ba2

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/organization/management/v1/OrganizationManagementSuccessTest.java [hunk 4]
```diff
@@ -111,9 +113,6 @@ public class OrganizationManagementSuccessTest extends OrganizationManagementBas
 
     protected OAuth2RestClient restClient;
 
-    private static final int NUM_OF_ORGANIZATIONS_FOR_PAGINATION_TESTS = 20;
```
Comment (ShanChathusanda93) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/organization/management/v1/OrganizationManagementSuccessTest.java:114]: Any specific reason to remove this constants?

Comment (BimsaraBodaragama) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/server/organization/management/v1/OrganizationManagementSuccessTest.java:114]: Moved it to the Base Class.

**Comment by BimsaraBodaragama** [comment]
All the tests passed locally:
image
_Source: target/surefire-reports/emailable-report.html_

**Review by sahandilshan — APPROVED** [review]
